src/ipld/frozen_cids.rs

```diff
@@ -0,0 +1,176 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use crate::utils::cid::CidVariant;
+use cid::Cid;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+/// Similar to the `CidHashMap` implementation, `FrozenCids` optimizes storage of
+/// CIDs that would normally be stored as a vector of CIDs. The `V1 DAG-CBOR Blake2b-256`
+/// variant (which can be stored in 32 bytes vs 96 bytes for a `Cid`
+/// type) is `+99.99%` of all CIDs, so very few CIDs need to be stored in the
+/// `Generic(Box<Cid>)` variant of `CidVariant`.
+///
+/// We use `Box<[...]>` to save memory, avoiding vector overallocation.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct FrozenCids(Box<[CidVariant]>);
+
+impl Default for FrozenCids {
+    fn default() -> Self {
+        FrozenCids(Box::new([]))
+    }
+}
+
+pub struct IntoIter<'a> {
+    cids: std::slice::Iter<'a, CidVariant>,
+}
+
+impl<'a> IntoIterator for &'a FrozenCids {
+    type Item = Cid;
+    type IntoIter = IntoIter<'a>;
+    fn into_iter(self) -> Self::IntoIter {
+        IntoIter {
+            cids: self.0.iter(),
+        }
+    }
+}
+
+impl<'a> Iterator for IntoIter<'a> {
+    type Item = Cid;
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.cids.next() {
+            Some(cid) => match cid {
+                CidVariant::V1DagCborBlake2b(bytes) => {
+                    Some(Cid::from(CidVariant::V1DagCborBlake2b(*bytes)))
+                }
+                CidVariant::Generic(cid) => Some(*cid.clone()),
+            },
+            None => None,
+        }
+    }
+}
+
+impl Serialize for FrozenCids {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        Vec::<Cid>::from(self).serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for FrozenCids {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        Ok(Self::from(<Vec<Cid>>::deserialize(deserializer)?))
+    }
+}
```

Comment:
I think you should implement `Serialize` and `Deserialize` for `CidVariant`, and then `#[derive(Serialize, Deserialize)]` for `FrozenCid`s

Reply:
Done.